cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,819 @@
+     -- changes (i.e. creations or deletions), or if the modification time and
+     -- content hash of any matching file has changed.
+     --
+   | MonitorFileGlob !FilePathGlob
+
+  deriving (Show, Generic)
+
+instance Binary MonitorFilePath
+
+-- | A file path specified by globbing
+--
+data FilePathGlob
+   = GlobDir  !Glob !FilePathGlob
+   | GlobFile !Glob
+  deriving (Eq, Show, Generic)
+
+instance Binary FilePathGlob
+
+-- | Creates a list of files to monitor when you search for a file which
+-- unsuccessfully looked in @notFoundAtPaths@ before finding it at
+-- @foundAtPath@.
+monitorFileSearchPath :: [FilePath] -> FilePath -> [MonitorFilePath]
+monitorFileSearchPath notFoundAtPaths foundAtPath =
+    MonitorFile foundAtPath
+  : map MonitorNonExistentFile notFoundAtPaths
+
+-- | Similar to 'monitorFileSearchPath', but also instructs us to
+-- monitor the hash of the found file.
+monitorFileHashedSearchPath :: [FilePath] -> FilePath -> [MonitorFilePath]
+monitorFileHashedSearchPath notFoundAtPaths foundAtPath =
+    MonitorFileHashed foundAtPath
+  : map MonitorNonExistentFile notFoundAtPaths
+
+
+------------------------------------------------------------------------------
+-- Implementation types, files status
+--
+
+-- | The state necessary to determine whether a set of monitored
+-- files has changed.  It consists of two parts: a set of specific
+-- files to be monitored (index by their path), and a list of
+-- globs, which monitor may files at once.
+data MonitorStateFileSet
+   = MonitorStateFileSet !(Map FilePath MonitorStateFile)
+                         ![FileGlobMonitorState]
+  deriving Show
+
+type Hash = Int
+#if MIN_VERSION_directory(1,2,0)
+type ModTime = UTCTime
+#else
+type ModTime = ClockTime
+#endif
+
+-- | The state necessary to determine whether a monitored file has changed.
+--
+-- This covers all the cases of 'MonitorFilePath' except for globs which is
+-- covered separately by 'FileGlobMonitorState'.
+--
+data MonitorStateFile
+   = MonitorStateFile       !ModTime         -- ^ cached file mtime
+   | MonitorStateFileHashed !ModTime !Hash   -- ^ cached file mtime and content hash
+   | MonitorStateFileNonExistent
+
+     -- | These two are to deal with the situation where we've been asked
+     -- to monitor a file that's expected to exist, but when we come to
+     -- check it's status, it no longer exists.
+   | MonitorStateFileChanged
+   | MonitorStateFileHashChanged
+  deriving (Show, Generic)
+
+instance Binary MonitorStateFile
+
+-- | The state necessary to determine whether the files matched by a globbing
+-- match have changed.
+--
+data FileGlobMonitorState
+   = MonitorStateGlobDirs  !Glob !FilePathGlob
+                           !ModTime
+                           ![(FilePath, FileGlobMonitorState)] -- invariant: sorted
+
+   | MonitorStateGlobFiles !Glob
+                           !ModTime
+                           ![(FilePath, ModTime, Hash)] -- invariant: sorted
```

Comment:
Similarly we don't have glob non-existent where as we do have that option for individual files.